src/signing.rs

```diff
@@ -0,0 +1,310 @@
+                server_signed.verify(server_public_key)?;
+                return Message::try_from(server_signed);
+            }
+        }
+        Err(Error::InvalidMessage)
+    }
+
+    pub fn sign_final<F: FnOnce() -> String>(
+        message_str: &str,
+        server_public_key: &impl PublicKeyTrait,
+        client_secret_key: &impl SecretKeyTrait,
+        password: F,
+    ) -> Result<OpenPGPMessage> {
+        let pgp_message = OpenPGPMessage::from_string(message_str)?.0;
+        pgp_message.verify(server_public_key)?;
+        let message = Message::try_from(pgp_message)?;
+        Ok(
+            OpenPGPMessage::Literal(LiteralData::from_str(&message.id.to_string(), message_str))
+                .sign(client_secret_key, password, HashAlgorithm::SHA2_256)?
+                .compress(CompressionAlgorithm::ZIP)?,
+        )
+    }
+}
+
+impl TryFrom<&Message> for OpenPGPMessage {
+    type Error = Error;
+
+    fn try_from(m: &Message) -> Result<Self, Self::Error> {
+        Ok(OpenPGPMessage::Literal(LiteralData::from_str(
+            &m.id.to_string(),
+            &serde_json::to_string(&m)?,
+        )))
+    }
+}
+
+impl TryFrom<Message> for OpenPGPMessage {
+    type Error = Error;
+
+    fn try_from(value: Message) -> Result<Self, Self::Error> {
+        Self::try_from(&value)
+    }
+}
+
+impl TryFrom<OpenPGPMessage> for Message {
+    type Error = Error;
+
+    fn try_from(value: OpenPGPMessage) -> Result<Self, Self::Error> {
+        Ok(serde_json::from_str(
+            &value
+                .decompress()?
+                .get_literal()
+                .map_or_else(|| Err(Error::InvalidMessage), Ok)?
+                .to_string()
+                .map_or_else(|| Err(Error::InvalidMessage), Ok)?,
+        )?)
+    }
+}
+
+// Unsafe, here for reference if for some reason this is needed (appears to work with `pgp = "0.7.1"`)
+// This was only thought of because the signature field in [`StandaloneSignature`] was private in the latest version of rpgp (at the time "0.7.1").
+//
+// Update:
+// Found a safe workaround:
+// ```rust
+// if let pgp::composed::message::Message::Signed {
+//     message,
+//     one_pass_signature: _,
+//     signature,
+// } = message
+// { /* use signature here */ }
+// ```
+//
+// Old unsafe workaround:
+// ```rust
+// pub trait GetSignature {
+//     fn get_signature(self) -> Signature;
+// }
+
+// impl GetSignature for StandaloneSignature {
+//     fn get_signature(self) -> Signature {
+//         struct StandaloneSignature {
+//             signature: Signature,
+//         }
+//         let transmuted: StandaloneSignature = unsafe { std::mem::transmute(self) };
+//         transmuted.signature
+//     }
+// }
+// ```
+
+pub async fn get_or_import_public_key(
+    fingerprint: &str,
+    key_server: &str,
+) -> Result<SignedPublicKey> {
+    let file_name = format!(
+        "{}{}.asc",
+        USER_PUBLIC_KEY_FOLDER,
+        fingerprint.to_ascii_uppercase()
+    );
```

Comment:
This isn't vulnerable to path traversal is it? Can fingerprint be user defined?

Reply:
Oh good point, I should double check that the given string is a valid fingerprint and also check that it has no slashes. Good catch